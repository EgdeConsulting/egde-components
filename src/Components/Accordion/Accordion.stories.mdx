import { Meta, Story, Canvas } from '@storybook/addon-docs/blocks';
import { Accordion } from './Accordion';
import { DocsPage } from 'Components';
import { LibPath } from 'Types';

export const Template = (args) => <Accordion {...args} />;

export const componentName = 'Accordion';
export const storyNames = [
    'Accordion OK',
    'Accordion InProgress',
    'More Accordion',
    'Only one Accordion open',
];

<DocsPage
    componentName={componentName}
    componentDescription="This component is used to display content in a vertical dropdown-like manner."
    addStories={true}
    storyNames={storyNames}
    addImportantNote={true}
    noteMessage="There is a bug with the 'Show Code' where nested components inside of props are being enclosed by a MDXCreateElement-tag."
    variant="stable"
/>

<Meta
    title={LibPath.Disclosure + componentName}
    component={Accordion}
    args={{
        items: [
            {
                title: 'Accordion 1',
                content: <div>Content 1</div>,
                badge: '',
            },
            {
                title: 'Accordion 2',
                content: <div>Content 2</div>,
                badge: '',
            },
        ],
        allowMultiple: true,
    }}
/>

<Canvas withToolbar>
    <Story name="Base">{Template.bind({})}</Story>
</Canvas>

<Canvas withToolbar>
    <Story
        name="Accordion Ok"
        args={{
            items: [
                {
                    title: 'Accordion 1',
                    content: <div>Content 1</div>,
                    badge: 'OK',
                },
                {
                    title: 'Accordion 2',
                    content: <div>Content 2</div>,
                    badge: 'OK',
                },
            ],
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

<Canvas withToolbar>
    <Story
        name="Accordion InProgress"
        args={{
            items: [
                {
                    title: 'Accordion 1',
                    content: <div>Content 1</div>,
                    badge: 'InProgress',
                },
                {
                    title: 'Accordion 2',
                    content: <div>Content 2</div>,
                    badge: 'InProgress',
                },
            ],
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

<Canvas withToolbar>
    <Story
        name="More Accordion"
        args={{
            items: [
                {
                    title: 'Accordion 1',
                    content: <div>Content 1</div>,
                    badge: '',
                },
                {
                    title: 'Accordion 2',
                    content: <div>Content 2</div>,
                    badge: '',
                },
                {
                    title: 'Accordion 3',
                    content: <div>Content 3</div>,
                    badge: '',
                },
                {
                    title: 'Accordion 4',
                    content: <div>Content 4</div>,
                    badge: '',
                },
                {
                    title: 'Accordion 5',
                    content: <div>Content 5</div>,
                    badge: '',
                },
                {
                    title: 'Accordion 6',
                    content: <div>Content 6</div>,
                    badge: '',
                },
            ],
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

<Canvas withToolbar>
    <Story
        name="Only one Accordion open"
        args={{
            allowMultiple: false,
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

<!--
<>
    {storiesOf('Components/Accordion'}, module).add('storiesOfTest', () => (
        <Accordion
            items={[
                {
                    title: 'Accordion 69',
                    content: <div>Content 69</div>,
                    badge: AccordionBadgeEnum.OK,
                },
                {
                    title: 'Accordion 420',
                    content: <div>Content 420</div>,
                    badge: AccordionBadgeEnum.OK,
                },
            ]}
            allowMultiple={true}
        />
    ))}
</>
-->
