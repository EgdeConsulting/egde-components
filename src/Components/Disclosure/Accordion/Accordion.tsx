import React, { ReactElement } from 'react';
import {
    Accordion as ChakraAccordion,
    AccordionItem,
    AccordionButton,
    AccordionPanel,
    AccordionIcon,
    Box,
    Badge,
    Stack,
    Text,
} from '@chakra-ui/react';
import { AccordionProps, AccordionBadgeEnum } from 'Types';

/*
    EXAMPLE USE:
    function AccordionExample(): ReactElement {
        const accordionItems = [
            {
                title: 'First accordion',
                content: <div>This is the first accordions content</div>,
            },
            {
                title: 'Second accordion',
                content: <div>This is the second accordions content</div>,
            },
            {
                title: 'Third accordion',
                content: <div>This is the third accordions content</div>,
            },
        ];
        return (
            <div>
                <Accordion items={accordionItems} allowMultiple />
            </div>
        );
    }
*/

const AccordionBadge = (props: {
    badgeType: AccordionBadgeEnum;
}): ReactElement => {
    const { badgeType } = props;
    const color =
        badgeType === AccordionBadgeEnum.OK ? 'badge.ok' : 'badge.wip';
    return (
        <span>
            <Badge
                background={color}
                color="black"
                variant="solid"
                fontWeight="thin"
                textTransform="none"
            >
                {badgeType}
            </Badge>
        </span>
    );
};

const Accordion = (props: AccordionProps): ReactElement => {
    const { items, allowMultiple } = props;

    return (
        <ChakraAccordion allowMultiple={allowMultiple}>
            {items.map((accordionItem, index) => {
                return (
                    <AccordionItem key={`aci${index}`}>
                        <h2>
                            <AccordionButton
                                _expanded={{ background: 'lightblue' }}
                            >
                                <>
                                    <Box flex="1" textAlign="left">
                                        <Stack spacing="0px" direction="column">
                                            <Text color="blue">
                                                {accordionItem.title}
                                            </Text>
                                            {accordionItem.badge && (
                                                <AccordionBadge
                                                    badgeType={
                                                        accordionItem.badge
                                                    }
                                                />
                                            )}
                                        </Stack>
                                    </Box>
                                    <AccordionIcon />
                                </>
                            </AccordionButton>
                        </h2>

                        <AccordionPanel pb={4}>
                            {accordionItem.content}
                        </AccordionPanel>
                    </AccordionItem>
                );
            })}
        </ChakraAccordion>
    );
};

export { Accordion };
