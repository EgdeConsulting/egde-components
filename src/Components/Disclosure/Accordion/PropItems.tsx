import React from 'react';

export const Default = [
    {
        title: 'Accordion 1',
        content: <div>Content 1</div>,
        badge: '',
    },
    {
        title: 'Accordion 2',
        content: <div>Content 2</div>,
        badge: '',
    },
];

export const AccordionOk = [
    {
        title: 'Accordion 1',
        content: <div>Content 1</div>,
        badge: 'OK',
    },
    {
        title: 'Accordion 2',
        content: <div>Content 2</div>,
        badge: 'OK',
    },
];

export const AccordionInProgress = [
    {
        title: 'Accordion 1',
        content: <div>Content 1</div>,
        badge: 'InProgress',
    },
    {
        title: 'Accordion 2',
        content: <div>Content 2</div>,
        badge: 'InProgress',
    },
];

export const MoreAccordion = [
    {
        title: 'Accordion 1',
        content: <div>Content 1</div>,
        badge: '',
    },
    {
        title: 'Accordion 2',
        content: <div>Content 2</div>,
        badge: '',
    },
    {
        title: 'Accordion 3',
        content: <div>Content 3</div>,
        badge: '',
    },
    {
        title: 'Accordion 4',
        content: <div>Content 4</div>,
        badge: '',
    },
    {
        title: 'Accordion 5',
        content: <div>Content 5</div>,
        badge: '',
    },
    {
        title: 'Accordion 6',
        content: <div>Content 6</div>,
        badge: '',
    },
];
