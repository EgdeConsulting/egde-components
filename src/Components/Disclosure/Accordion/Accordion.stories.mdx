import { Meta, Story, Canvas } from '@storybook/addon-docs';
import { DocsPageHeader, DocsPageBodyComponents } from 'Docs';
import { LibPath } from 'Types';
import { Accordion } from 'Components';
import * as PropItems from './PropItems';

export const Template = (args) => <Accordion {...args} />;

export const componentName = 'Accordion';

<Meta
    title={LibPath.Disclosure + componentName}
    component={Accordion}
    args={{
        items: PropItems.Default,
        allowMultiple: true,
    }}
/>

<DocsPageHeader
    title={componentName}
    description="This component is used to display content in a vertical dropdown-like manner."
    addImportantNote={false}
    variant="stable"
/>

<Canvas withToolbar>
    <Story name="Base">{Template.bind({})}</Story>
</Canvas>

<DocsPageBodyComponents addStories={true} />

### Accordion Ok

<Canvas withToolbar>
    <Story
        name="Accordion Ok"
        args={{
            items: PropItems.AccordionOk,
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Accordion InProgress

<Canvas withToolbar>
    <Story
        name="Accordion InProgress"
        args={{
            items: PropItems.AccordionInProgress,
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### More Accordion

<Canvas withToolbar>
    <Story
        name="More Accordion"
        args={{
            items: PropItems.MoreAccordion,
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Only one Accordion open

<Canvas withToolbar>
    <Story
        name="Only one Accordion open"
        args={{
            allowMultiple: false,
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>
