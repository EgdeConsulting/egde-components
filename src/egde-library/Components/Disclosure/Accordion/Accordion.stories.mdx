import { Meta, Story, Canvas } from '@storybook/addon-docs';
import { LibPath, DocsPageHeader, DocsPageBodyComponents } from 'egde-library';
import { Accordion } from 'egde-components';
import * as PropItems from './PropItems';

export const Template = (args) => <Accordion {...args} />;

export const componentName = 'Accordion';

<Meta
    title={LibPath.Disclosure + componentName}
    component={Accordion}
    args={{
        items: PropItems.Default,
        allowMultiple: true,
    }}
/>

<DocsPageHeader
    title={componentName}
    description="This component is used to display content in a vertical dropdown-like manner."
    addImportantNote={false}
    variant="stable"
/>

<Canvas withToolbar>
    <Story name="Base">{Template.bind({})}</Story>
</Canvas>

<DocsPageBodyComponents addStories={true} />

### Accordion Ok

<Canvas withToolbar>
    <Story
        name="Accordion Ok"
        args={{
            items: PropItems.AccordionOk,
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Accordion InProgress

<Canvas withToolbar>
    <Story
        name="Accordion InProgress"
        args={{
            items: PropItems.AccordionInProgress,
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### More Accordion

<Canvas withToolbar>
    <Story
        name="More Accordion"
        args={{
            items: PropItems.MoreAccordion,
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

### Only one Accordion open

<Canvas withToolbar>
    <Story
        name="Only one Accordion open"
        args={{
            allowMultiple: false,
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>
