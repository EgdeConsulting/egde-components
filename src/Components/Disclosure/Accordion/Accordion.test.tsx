import React, { ReactElement }  from 'react';
import '@testing-library/jest-dom';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Accordion } from './Accordion';


const ACCORDION_TITLES: string[] = ['First', 'Second', 'Third'];
const ACCORDION_CONTENT: string[] = ['FirstContent', 'SecondContent', 'ThirdContent'];

const AccordionTester = (): ReactElement =>{
    const accordionItems = [
        {
            title: ACCORDION_TITLES[0],
            content: <div>{ACCORDION_CONTENT[0]}</div>,
        },
        {
            title: ACCORDION_TITLES[1],
            content: <div>{ACCORDION_CONTENT[1]}</div>,
        },
    ];
    return <Accordion items={accordionItems} allowMultiple />;
}


describe('Accordion component test', () => {
  it('should render correctly', () => {
    const tree = render(<AccordionTester />);
    expect(tree).toMatchSnapshot();
  });

    it('should render accordions unexpanded, and show no content', () => {
        render(<AccordionTester />);
        const firstAccordion = screen.getByRole('button', {
            name: ACCORDION_TITLES[0],
        });
        const secondAccordion = screen.getByRole('button', {
            name: ACCORDION_TITLES[1],
        });
        expect(firstAccordion).toBeVisible();
        expect(secondAccordion).toBeVisible();
        expect(firstAccordion).toHaveAttribute('aria-expanded', 'false');
        expect(secondAccordion).toHaveAttribute('aria-expanded', 'false');
        expect(screen.getByText(ACCORDION_CONTENT[0])).not.toBeVisible();
        expect(screen.getByText(ACCORDION_CONTENT[1])).not.toBeVisible();
    });
    it('should show content when accordion is clicked, and hide it when clicked again', async () => {
        render(<AccordionTester />);
        const firstAccordion = screen.getByRole('button', {
            name: ACCORDION_TITLES[0],
        });
        userEvent.click(firstAccordion);
        await waitFor(() => {
            expect(screen.getByText(ACCORDION_CONTENT[0])).toBeVisible();
        });
        await waitFor(() => {
            expect(firstAccordion).toHaveAttribute('aria-expanded', 'true');
        });
        userEvent.click(firstAccordion);
        await waitFor(() => {
            expect(screen.getByText(ACCORDION_CONTENT[0])).not.toBeVisible();
        });
        await waitFor(() => {
            expect(firstAccordion).toHaveAttribute('aria-expanded', 'false');
        });
    });
    it('should show all content when all accordions are clicked', async () => {
        render(<AccordionTester />);
        const firstAccordion = screen.getByRole('button', {
            name: ACCORDION_TITLES[0],
        });
        const secondAccordion = screen.getByRole('button', {
            name: ACCORDION_TITLES[1],
        });
        userEvent.click(firstAccordion);
        await waitFor(() => {
            expect(screen.getByText(ACCORDION_CONTENT[0])).toBeVisible();
        });
        userEvent.click(secondAccordion);
        await waitFor(() => {
            expect(screen.getByText(ACCORDION_CONTENT[1])).toBeVisible();
        });
    });
});
