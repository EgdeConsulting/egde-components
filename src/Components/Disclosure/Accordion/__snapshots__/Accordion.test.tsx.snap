// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component test matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
    <div>
      <div
        class="chakra-accordion css-0"
      >
        <div
          class="chakra-accordion__item css-1fsnuue"
        >
          <h2>
            <button
              aria-controls="accordion-panel-1"
              aria-expanded="false"
              class="chakra-accordion__button css-o0n1ae"
              data-index="0"
              id="accordion-button-1"
              type="button"
            >
              <div
                class="css-1eziwv"
              >
                <div
                  class="chakra-stack css-1h3nfwc"
                >
                  <p
                    class="chakra-text css-14ksm7b"
                  >
                    First
                  </p>
                </div>
              </div>
              <svg
                aria-hidden="true"
                class="chakra-icon chakra-accordion__icon css-8o5ky2"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </h2>
          <div
            class="chakra-collapse"
            style="overflow: hidden; display: none; opacity: 0; height: 0px;"
          >
            <div
              aria-labelledby="accordion-button-1"
              class="chakra-accordion__panel css-b0qdn7"
              id="accordion-panel-1"
              role="region"
            >
              <div>
                FirstContent
              </div>
            </div>
          </div>
        </div>
        <div
          class="chakra-accordion__item css-1fsnuue"
        >
          <h2>
            <button
              aria-controls="accordion-panel-2"
              aria-expanded="false"
              class="chakra-accordion__button css-o0n1ae"
              data-index="1"
              id="accordion-button-2"
              type="button"
            >
              <div
                class="css-1eziwv"
              >
                <div
                  class="chakra-stack css-1h3nfwc"
                >
                  <p
                    class="chakra-text css-14ksm7b"
                  >
                    Second
                  </p>
                </div>
              </div>
              <svg
                aria-hidden="true"
                class="chakra-icon chakra-accordion__icon css-8o5ky2"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </h2>
          <div
            class="chakra-collapse"
            style="overflow: hidden; display: none; opacity: 0; height: 0px;"
          >
            <div
              aria-labelledby="accordion-button-2"
              class="chakra-accordion__panel css-b0qdn7"
              id="accordion-panel-2"
              role="region"
            >
              <div>
                SecondContent
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-accordion css-0"
    >
      <div
        class="chakra-accordion__item css-1fsnuue"
      >
        <h2>
          <button
            aria-controls="accordion-panel-1"
            aria-expanded="false"
            class="chakra-accordion__button css-o0n1ae"
            data-index="0"
            id="accordion-button-1"
            type="button"
          >
            <div
              class="css-1eziwv"
            >
              <div
                class="chakra-stack css-1h3nfwc"
              >
                <p
                  class="chakra-text css-14ksm7b"
                >
                  First
                </p>
              </div>
            </div>
            <svg
              aria-hidden="true"
              class="chakra-icon chakra-accordion__icon css-8o5ky2"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </button>
        </h2>
        <div
          class="chakra-collapse"
          style="overflow: hidden; display: none; opacity: 0; height: 0px;"
        >
          <div
            aria-labelledby="accordion-button-1"
            class="chakra-accordion__panel css-b0qdn7"
            id="accordion-panel-1"
            role="region"
          >
            <div>
              FirstContent
            </div>
          </div>
        </div>
      </div>
      <div
        class="chakra-accordion__item css-1fsnuue"
      >
        <h2>
          <button
            aria-controls="accordion-panel-2"
            aria-expanded="false"
            class="chakra-accordion__button css-o0n1ae"
            data-index="1"
            id="accordion-button-2"
            type="button"
          >
            <div
              class="css-1eziwv"
            >
              <div
                class="chakra-stack css-1h3nfwc"
              >
                <p
                  class="chakra-text css-14ksm7b"
                >
                  Second
                </p>
              </div>
            </div>
            <svg
              aria-hidden="true"
              class="chakra-icon chakra-accordion__icon css-8o5ky2"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </button>
        </h2>
        <div
          class="chakra-collapse"
          style="overflow: hidden; display: none; opacity: 0; height: 0px;"
        >
          <div
            aria-labelledby="accordion-button-2"
            class="chakra-accordion__panel css-b0qdn7"
            id="accordion-panel-2"
            role="region"
          >
            <div>
              SecondContent
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
